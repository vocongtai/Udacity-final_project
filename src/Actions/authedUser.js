export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const SET_LOGOUT_USER = "SET_LOGOUT_USER";

export function setAuthedUser(authedUser) {
  return {
    type: SET_AUTHED_USER,
    authedUser,
  };
}

export function setLogoutUser() {
  return {
    type: SET_LOGOUT_USER,
  };
}

//login handle func action
export function logginHandle(username, password) {
  return (dispatch, getState) => {
    const { users } = getState();
    const user = Object.values(users).find(
      (user) => user.id === username && user.password === password
    );
    if (!!user) {
      dispatch(setAuthedUser(user));
    }
  };
}

//logout handle func action
export function logoutHandle() {
  return (dispatch) => {
    return dispatch(setLogoutUser());
  };
}
