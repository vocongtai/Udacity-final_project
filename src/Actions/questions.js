import { saveQuestion, saveQuestionAnswer } from "../Utils/Api";
import { addQuestionUser, addAnswerUser } from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";

export function receiverQuestion(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function addAnswerQuestion(author, qid, answer) {
  return {
    type: ADD_ANSWER_QUESTION,
    author,
    qid,
    answer,
  };
}

export function handleAddQuestion(firstOption, secondOption) {
  return (dispatch, getState) => {
      const { authedUser } = getState();

      return saveQuestion(firstOption, secondOption, authedUser)
          .then((question) => {
              dispatch(addQuestion(question));
              dispatch(addQuestionUser(question))
          })
  };
}

export function handleAddAnswerQuestion(qid, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    return saveQuestionAnswer(authedUser.id, qid, answer).then(() => {
      dispatch(addAnswerUser(authedUser.id, qid, answer));
      dispatch(addAnswerQuestion(authedUser.id, qid, answer));
    });
  };
}
