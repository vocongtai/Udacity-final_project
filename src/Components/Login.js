import React, { useState } from "react";
import { connect } from "react-redux";
import { Navigate } from "react-router-dom";
import {logginHandle} from "../Actions/authedUser"

const Login = ({ dispatch, userLogin }) => {
  const [userName, setUserName] = useState("tylermcginnis");
  const [passWord, setPassword] = useState("abc321");

  if(userLogin){
    const params=new URLSearchParams(window.location.search);
    const redirectURL=params.get('redirectTo');
    return <Navigate to={redirectURL ?redirectURL : "/"}/>
  }

  const userNameHandle = (event) => {
    //
    const username = event.target.value;
    setUserName(username);
  };

  const passwordHandle = (event) => {
    //
    const password = event.target.value;
    setPassword(password);
  };

  const loginHandle = (event) => {
    //
    event.preventDefault();
    dispatch(logginHandle(userName,passWord))

    setUserName("");
    setPassword("");
  };

  return (
    <div className="login-container">
        <h1 className="center" data-testid="login-header">Welcome to Poll Page App! </h1>
      <form onSubmit={loginHandle} className="login-form">
        <label className="center" data-testid="userName-label">Username: </label>
        <input
          type="text"
          placeholder="Input Username ...."
          name="userName"
          id="userName"
          onChange={userNameHandle}
          value={userName}
          data-testid="username-input"
        />
        <label className="center" data-testid="password-label">Password: </label>
        <input
          type="password"
          placeholder="***"
          name="password"
          id="password"
          onChange={passwordHandle}
          value={passWord}
          data-testid="password-input"
        />
        <button className="btn-login" data-testid="login-btn">Login</button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => {
  return {
    userLogin: !!authedUser,
  };
};

export default connect(mapStateToProps)(Login);
