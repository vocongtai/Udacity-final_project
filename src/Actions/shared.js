import { getInitialData } from "../Utils/Api";
import { receiverQuestion } from "./questions";
import { receiverUsers } from "./users";

export function handleInitialData() {
  return (dispatch) => {
    return getInitialData().then(({ users, questions }) => {
      dispatch(receiverQuestion(questions));
      dispatch(receiverUsers(users));
    });
  };
}
