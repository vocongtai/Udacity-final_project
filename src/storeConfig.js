import { configureStore } from "@reduxjs/toolkit";
import authedUser from "./reducers/authedUser";
import users from "./reducers/users";
import questions from "./reducers/questions";

export const storeConfig = configureStore({
  reducer: {
    authedUser,
    users,
    questions,
  },
});
