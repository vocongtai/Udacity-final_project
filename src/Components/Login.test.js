import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { storeConfig } from "../storeConfig";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import Login from "./Login";
import { handleInitialData } from "../Actions/shared";

describe("Login", () => {
  it("Should render the login component", () => {
    const componentLogin = render(
      <Provider store={storeConfig}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    expect(componentLogin).toBeDefined();
    expect(componentLogin).toMatchSnapshot();
  });

  it("Should set empty text when input wrong password", () => {
    const componentLogin = render(
      <Provider store={storeConfig}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    const loginHeader = componentLogin.getByTestId("login-header");
    const usernameLabel = componentLogin.getByTestId("userName-label");
    const usernameInput = componentLogin.getByTestId("username-input");
    const passwordLabel = componentLogin.getByTestId("password-label");
    const passwordInput = componentLogin.getByTestId("password-input");
    const loginBtn = componentLogin.getByTestId("login-btn");

    expect(loginHeader).toBeInTheDocument();
    expect(usernameLabel).toBeInTheDocument();
    expect(usernameInput).toBeInTheDocument();
    expect(passwordLabel).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();

    fireEvent.change(usernameInput, { target: { value: "sarahedo" } });
    fireEvent.change(passwordInput, { target: { value: "testwrongpassword" } });
    expect(usernameInput.value).toBe("sarahedo");
    expect(passwordInput.value).toBe("testwrongpassword");
    fireEvent.click(loginBtn);
    expect(loginHeader).toBeInTheDocument();
    expect(usernameInput.value).toBe("");
    expect(passwordInput.value).toBe("");
  });
});
