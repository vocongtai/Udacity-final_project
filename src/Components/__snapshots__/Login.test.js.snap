// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Should render the login component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
      >
        <h1
          class="center"
          data-testid="login-header"
        >
          Welcome to Poll Page App! 
        </h1>
        <form
          class="login-form"
        >
          <label
            class="center"
            data-testid="userName-label"
          >
            Username: 
          </label>
          <input
            data-testid="username-input"
            id="userName"
            name="userName"
            placeholder="Input Username ...."
            type="text"
            value="tylermcginnis"
          />
          <label
            class="center"
            data-testid="password-label"
          >
            Password: 
          </label>
          <input
            data-testid="password-input"
            id="password"
            name="password"
            placeholder="***"
            type="password"
            value="abc321"
          />
          <button
            class="btn-login"
            data-testid="login-btn"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
    >
      <h1
        class="center"
        data-testid="login-header"
      >
        Welcome to Poll Page App! 
      </h1>
      <form
        class="login-form"
      >
        <label
          class="center"
          data-testid="userName-label"
        >
          Username: 
        </label>
        <input
          data-testid="username-input"
          id="userName"
          name="userName"
          placeholder="Input Username ...."
          type="text"
          value="tylermcginnis"
        />
        <label
          class="center"
          data-testid="password-label"
        >
          Password: 
        </label>
        <input
          data-testid="password-input"
          id="password"
          name="password"
          placeholder="***"
          type="password"
          value="abc321"
        />
        <button
          class="btn-login"
          data-testid="login-btn"
        >
          Login
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
